src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java

```diff
@@ -39,30 +47,35 @@ public SSLEngineResult wrap(final ByteBuffer[] srcs, final int offset, final int
 
     @Override
     public SSLEngineResult unwrap(final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
-        final SSLEngineResult result;
-        if (generatedJa3Signature) {
-            result = engine.unwrap(src, dsts, offset, length);
-        } else {
-            // Ensure no-one modifies the buffer underneath as per sun.security.ssl.SSLEngineImpl
-            String ja3 = null;
-            final HandshakeStatus handshakeStatus = engine.getHandshakeStatus();
-            if (HandshakeStatus.NOT_HANDSHAKING != handshakeStatus && HandshakeStatus.FINISHED != handshakeStatus) {
-                ja3 = new JA3Signature().ja3Signature(src);
-            } else {
-                generatedJa3Signature = true;
-            }
-
-            result = engine.unwrap(src, dsts, offset, length);
-
-            if (ja3 != null) {
-                // no lock needed, it is fine if we write same digest multiple times
-                engine.getSession().putValue(JA3Constants.JA3_FINGERPRINT, ja3);
-                generatedJa3Signature = true;
-            }
-        }
-
-
-        return result;
+		final SSLEngineResult result;
+		if (ja3 == null) {
+			//1. Generate JA3 signature
+			final HandshakeStatus handshakeStatus = engine.getHandshakeStatus();
+			if (HandshakeStatus.FINISHED != handshakeStatus) {
+				ja3 = new JA3Signature().ja3Signature(src);
+			}
+			
+			//unwrap
+			result = engine.unwrap(src, dsts, offset, length);
+		} else if (!isJa3SignatureSet) {
+			//2. Set ja3 signature in ssl handshake session
+			
+			//unwrap
+			result = engine.unwrap(src, dsts, offset, length);
+			
+			final SSLSession handshakeSession = engine.getHandshakeSession();
+			if (handshakeSession != null) {
+				//set ja3 signature in handshake session
+				handshakeSession.putValue(JA3Constants.JA3_FINGERPRINT, ja3);
+				isJa3SignatureSet = true;
+			}
+		} else {
```

Comment:
I haven't fully reviewed this code. The idea is to make the flows to unwrap post handshake fast with as many less checks as possible. The handshake + ja3 generation can be somewhat expensive and still efficient. Can you review the code and see?